docs/lesson-plans/week-5.md

```diff
@@ -0,0 +1,639 @@
+    
+    // Our new externs.
+    Externals externs;
+
+    // The basic constructor.
+    public PostsRouter(IPostsProcessor processor) {
+        this.processor = processor;
+        this.externs = new Externals();
+    }
+
+    // The override constructor.
+    public PostsRouter(IPostsProcessor processor, Externals override) {
+        this.processor = processor;
+        this.externs = override;
+    }
+
+    public static class Externals {
+        public Router getRouter(Vertx vertx) {
+            return Router.router(vertx);
+        }
+    }
+
+    @Override
+    public Router initializeRouter(Vertx vertx) {
+        Router router = externs.getRouter(vertx);
+    ...
+```
+
+And this is how we override that behavior in the test class:
+```java
+public class PostsRouterTest {
+    ...
+
+    // Create the new externals, which overides the old behavior.
+    public static class TestExterns extends Externals {
+        public Router getRouter(Vertx vertx) {
+            return myTestRouter;
+        }
+    }
+
+    @BeforeEach
+    public void setup() {
+        ...
+        this.processor = new Processor(this.db, new TestExterns());
+        ...
+    }
+}
+```
+
+#### Exceptions
+
+When testing exceptions, JUnit provides methods and annotations to help, but we find it more helpful to use the 
+try/catch way of doing things. We think it allows you to test things about the exception better.
+
+Example:
+```java 
+try {
+    thingThatThrowsException();
+    fail(); // fail this test if we make it this far
+}
+catch (TheExpectedExceptionType e) {
+    assertEquals("My exception message", e.getMessage());
+    assertEquals(otherThingsAboutExceptionIfYouWant, e.doSomething());
+}
+```
+
+### Testing Using Mockito
+
+Mockito is a really useful library for mocking the interactions between objects that you don't have control over
+(like dependencies). It lets you create fake versions of objects which you have complete control over, and it even lets
+you verify things about how it was called.
+
+If you're interested in learning more about Mockito, [here is the 
+documentation](http://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html). It even lets you do some 
+really cool things like `spy()`, where you can pass in an existing object, have it do _most_ of the things it usually 
+does, and then just override/verify/mock what you want!
+
+#### Mocking an Object and Overriding Returns
+
+To mock an object, all you have to do is call (the static method) `mock(Thing.class)`, which will return a fake
+instance of that `Thing` class/interface. By default, it will return null 
+for every non-void method you call, but you can override that using `thenReturn`, `thenThrow`, `then` (`thenAnswer`, 
+allows you to do custom stuff), or `doReturn`, `doThrow`, `doAnswer`, `doNothing`, `doCallRealMethod` in the case of 
+void returns. You can also chain them together to have it return/throw/do things one after the other in the order 
```

Comment:
You can also chain these then/do methods together to have the instance return/throw/do each thing in order each time the overridden method is called.